### ISSUE TITLE
clicking on any link within *org-roam* buffer fails with an error message

### ISSUE THREAD

**Issue by emacs18** [org-roam/org-roam#1732]
After upgrading to org-roam version 2, none of the backlinks shown in my `*org-roam*` buffer work.  If I click on any link  with my mouse or hit `RETURN` key with cursor on a link, nothing happens other than the following error being displayed in the mini-buffer:
```
There is no thing at point that could be visited
```
When I click on any link within `*org-roam*` buffer, I expected the the content of the target link to pop up and be displayed in a buffer.  However nothing happens others than the error message being displayed in the mini-buffer.

I have many emacs startup files setup using chemacs2 package on ubuntu 20.04 so that I can choose one of many setups.  The reported problem occurs with my recent builds of emacs 27 or 28, with package.el or straight.el, with spacemacs or no spacemacs.

**Comment by ayrat555** [comment]
https://github.com/org-roam/org-roam/issues/1732

**Comment by emacs18** [comment]
I just tried doom emacs.  The backlinks work for both emacs 27 and 28 in that clicking on a link brings up the corresponding buffer.  I guess I'll be comparing org-roam setup within doom emacs and spacemacs to see what the difference is.

**Comment by emacs18** [comment]
Doom works for me, and spacemacs does not.  One immediate difference is that `*org-roam*` buffer contents are different.  In doom `Backlinks` line is at the top, but it is second from bottom in spacemacs. Also in doom hitting `TAB` with cursor on `Backlinks` line toggles between collapsed vew and expanded view in similar fashion as in magit buffers.  In spacemacs, hitting `TAB` results in `Cannot hide root section` error message only.

**Comment by emacs18** [comment]
If I use the following as my `init.el` file, then backlinks work:
```elisp
(require 'package)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(package-initialize)
(package-refresh-contents)

(unless (package-installed-p 'use-package)
  (package-install 'use-package))

(require 'use-package)
(require 'org)

(use-package org-roam
  :ensure t
  :bind (("C-c n l" . org-roam-buffer-toggle)
         ("C-c n f" . org-roam-node-find)
         ("C-c n g" . org-roam-graph)
         ("C-c n i" . org-roam-node-insert)
         ("C-c n c" . org-roam-capture)
         ;; Dailies
         ("C-c n j" . org-roam-dailies-capture-today))
  :config
  (setq org-roam-v2-ack t)
  (setq org-roam-directory (expand-file-name user-emacs-directory))
  (org-roam-setup)
  (require 'org-roam-protocol))

(use-package helm :ensure t :config (helm-mode))
```
So this may be a spacemacs issue rather than org-roam issue.

**Comment by ayrat555** [comment]
I use vanilla emacs, but still, links can not be visited

**Comment by jethrokuan** [comment]
Disable page-break-lines-mode for now.

**Comment by emacs18** [comment]
I used "vanilla" emacs 27 or 28 as well.  Code I shared above was my `init.el`.  With this `*org-roam*` buffer displayed lines similar to doom emacs in that the first line was `Backlinks (1)`, hitting `TAB` with cursor on this line collapsed or expanded other lines in the buffer (as is done within magit mode sections), and hitting `RETURN` with cursor on links popped up appropriate buffers.  In short the backlink buffer seem to work just fine.

**Comment by emacs18** [comment]
Disabling `page-break-lines-mode` fixed the problem even on my spacemacs.
Thanks.

**Comment by dalanicolai** [comment]
So the above link is to the related and informative bug report at Spacemacs github

**Comment by shivamvats** [comment]
I use Spacemacs but disabling `page-break-lines-mode` does not work for me. However, `org-roam-buffer-refresh` fixes the buffer.

**Comment by mattford63** [comment]
org-roam-buffer-refresh only fixes the buffer if page-break-lines-mode is disabled

**Comment by VictorYYW** [comment]
Confirming that org-roam-buffer-refresh does fix the issue for me on spacemacs. Do we have to call it every time we switch to a new node? Would it be possible to set a hook for that?

**Comment by emacs18** [comment]
Ever since I disabled `page-break-lines-mode` a month ago, I've had no problem with orb-roam buffer.  I never have to call org-roam-buffer-refresh.  I use latest develop branch of spacemacs.

**Comment by lazzalazza** [comment]
The problem is still there form me (on  Spacemacs). I have **disabled global-page-break-lines mode** but that didn't affect the unwanted behaviour in any way. I still have to refresh the roam backlinks buffer every single time... Is there a fix for this?

**Comment by swang251** [comment]
As stated in issue #1764, the problem is related to `dotspacemacs-line-numbers` as well. 

The problem should be the `spacemacs` instead of `org-roam`, see discussion in #1943.

**Comment by ning-y** [comment]
@swang251 I have this error on doom emacs, so I do not think it is spacemacs specific. Unless doom is built on-top of spacemacs or vice versa (do correct me if I'm wrong).

**Comment by swang251** [comment]
Hi @ning-y , you could be right. There seem to be two modes that can potentially cause the problem, which is `page-break-lines-mode` and `display-line-numbers-mode`. 

I've only tested the `display-line-numbers-mode` with `org-roam` in the emacs with a minimum configuration and there is no problem at all. The problem can be triggered by adding `(add-hook 'org-mode-hook 'display-line-numbers-mode)` in the spacemacs configuration. Adding the same snippet in plain emacs doesn't trigger the problem.

**Comment by ribelo** [comment]
For some time now, an identical problem has occurred in `doom-emacs`. 

Opening a link works the first time the buffer shows up, or after a refresh, otherwise `<ret>` starts behaving like `evil-ret`, and `org-roam-preview-visit` returns a `no file at point` message .

**Comment by meliache** [comment]
Disabling `page-break-lines` for `org-mode` solved the issue for me. Just disabling it in the org-roam-buffer (i.e. for `org-roam-mode`) didn't solve the issue, weirdly :shrug:

To do that, I tried `page-break-lines-mode-maybe` as a hook, which only enables it for modes deriving from the modes in `page-break-lines-modes`. However, I ran into the limitation that I want to enable it for `outline-mode`, as this is what emacs NEWS use (`C-h n`), but `org-mode` derives for that. So I use my own function which also checks a deny-list:
``` elisp
  (setq meliache-page-break-lines-deny-modes '(org-mode))

  (defun meliache-page-break-lines-mode-maybe ()
    "Like `page-break-lines-mode-maybe', but in addition to the
allow-list `page-break-lines-modes' check also the deny-list `meliache-page-break-lines-deny-modes'.

. The issue is that I want to allow page-break-lines-mode in some
modes, but I want to disallow it in some of their derived modes,
e.g. org-roam-mode, where it leads to links not working
https://github.com/org-roam/org-roam/issues/1732. "
    (when (not (apply #'derived-mode-p meliache-page-break-lines-deny-modes))
      (page-break-lines-mode-maybe)))
```

**Comment by Sergey-Makarov** [comment]
Hey guys, here is a one-liner that fixes the issue for me:
```
(advice-add #'org-roam-fontify-like-in-org-mode :around (lambda (fn &rest args) (save-excursion (apply fn args))))
```

I am not very experienced in Emacs Lisp and have no idea why `org-roam-fontify-like-in-org-mode` changes the point inside the org-roam buffer, though 🤷

I am using Spacemacs, "global-page-break-lines-mode" and "global-display-line-numbers-mode" are enabled. The corresponding minor modes are enabled in my org mode buffer as well. Disabling all of those modes also makes the problem go away.

Org roam version: 20221231.2122

**Comment by real-or-random** [comment]
> Hey guys, here is a one-liner that fixes the issue for me:
> 
> ```
> (advice-add #'org-roam-fontify-like-in-org-mode :around (lambda (fn &rest args) (save-excursion (apply fn args))))
> ```
> 

Nice, that works also for me. This means we know how to fix this bug here. Are you willing to open a PR that adds `save-excursion` to `org-roam-fontify-like-in-org-mode`?

**Comment by Sergey-Makarov** [comment]
Will open a PR soon (in a few days maybe).

Initially I did not want push this solution as it is, because it looks more like a hack rather that something that addresses the cause directly and in a self-explanatory way. But it seems there is nothing else we can do about it.

**Comment by dankessler** [comment]
After stepping through the code *very* carefully with edebug, I think I know *what's* happening and how to fix it, although I don't fully understand *why* the problem is happening.

TLDR: The weird behavior under `page-break-mode` (present issue and https://github.com/syl20bnr/spacemacs/issues/14969) and `dotspacemacs-line-numbers` (#1764) can be fixed by wrapping the body of [`org-roam-fontify-like-in-org-mode`](https://github.com/org-roam/org-roam/blob/8667e441876cd2583fbf7282a65796ea149f0e5f/org-roam-utils.el#L209-L234) in a `save-excursion` form.

Side note: `org-roam-fontify-like-in-org-mode` seems like it's a hacked version of `org-fontify-like-in-org-mode` wrapped in a `let` form to set `org-ref-buffer-hacked` to `t`, but `org-ref-buffer-hacked` has [been removed](https://github.com/jkitchin/org-ref/commit/c13e33f) from `org-ref` anyway, so we should probably just call `org-fontify-like-in-org-mode` directly. It'd be nice to get the `save-excursion` wrapping into this function upstream, but in the meantime we could define a helper function that just wraps it in such a form.

Explanation: The call stack is something like
1. `org-roam-buffer-toggle`
2. `org-roam-buffer-persistent-redisplay`
3. `org-roam-buffer-render-contents` (wrapped in a `with-current-buffer`)
4. `org-roam-backlinks-section` (assuming the default value of `org-roam-mode-sections`
5. `org-roam-node-insert-section`
6. `org-roam-fontify-like-in-org-mode`

If we zoom in, here is the [most relevant bit](https://github.com/org-roam/org-roam/blob/8667e441876cd2583fbf7282a65796ea149f0e5f/org-roam-mode.el#L406-L408)

```
    (magit-insert-section section (org-roam-preview-section)
      (insert (org-roam-fontify-like-in-org-mode
               (org-roam-preview-get-contents (org-roam-node-file source-node) point))
              "\n")
```

When I stepped through this code, before it ran I evaluated `(point)` and it returned something at the end of the buffer (the under-construction `*org-roam*` buffer). Right after `org-roam-fontify-like-in-org-mode` returns its fontified string, `(point)` is suddenly at the beginning of the buffer, and thus new content gets inserted at the top rather than the end, and then everything is all mucked up. 

org-roam-fontify-like-in-org-mode` is currently wrapped inside a `with-temp-buffer` form. For some reason I don't fully grok, page-break-mode or `dotspacemacs-line-numbers` causes the value of `(point)` to not be preserved after exiting this `with-temp-buffer` form, but if both of those are off, it's preserved.

The preservation of point seems kind of fragile, e.g., if I set things up so that the org-roam buffer looks nice, but then I use edebug to step into the fontification code and step back out, point is not preserved.

Under the hood, `with-temp-buffer` uses `with-current-buffer` which in turn uses `save-current-buffer`. The documentation for `save-excursion` advises that it will save current buffer *and point*, but that if you only care about the buffer you should use `save-current-buffer` instead.

**Comment by real-or-random** [comment]
It would be good to know if anyone on Spacemacs can still reproduce this. @Sergey-Makarov noted in https://github.com/org-roam/org-roam/pull/2334 that he can't reproduce it, but he's no longer using Spacemacs. I also switched to doom, so I can't test this easily either.

**Comment by hpfr** [comment]
Based on Sergey’s recent feedback, let’s close this for now and revisit it if anyone pipes up with steps to reproduce.
